src/edu/csus/ecs/pc2/core/execute/Executable.java

```diff
@@ -2524,10 +2525,35 @@ public String substituteAllStrings(Run inRun, String origString) {
                 newString = replaceString(newString, "{:exitvalue}", Integer.toString(executionData.getExecuteExitValue()));
                 newString = replaceString(newString, "{:executetime}", Long.toString(executionData.getExecuteTimeMS()));
             }
+            
             String pc2home = new VersionInfo().locateHome();
             if (pc2home != null && pc2home.length() > 0) {
                 newString = replaceString(newString, "{:pc2home}", pc2home);
             }
+            
+            // TODO add to logic to use  problem.getSandboxType() and get clarification
+            // on how to handle sandbox variables if getSandboxType() == NONE, and other requirements
+            
+            // TODO REFACTOR replace vars with constants for: memlimit, sandboxcommandline,sandboxprogramname
+            newString = replaceString(newString, "{:memlimit}", Integer.toString(problem.getMemoryLimitMB()));
+
+            if (newString.indexOf("{:sandboxcommandline}") > -1) {
+                if (!StringUtilities.isEmpty(problem.getSandboxCmdLine())) {
+
+                    newString = replaceString(newString, "{:sandboxcommandline}", problem.getSandboxCmdLine());
+                    newString = substituteAllStrings(inRun, newString);
+                } else {
+                    newString = replaceString(newString, "{:sandboxcommandline}", "");
+                }
+            }
+
+            if (!StringUtilities.isEmpty(problem.getSandboxProgramName())) {
+                newString = replaceString(newString, "{:sandboxprogramname}", problem.getSandboxProgramName());
+            } else {
+                newString = replaceString(newString, "{:sandboxprogramname}", "");
+            }
+             
```

Comment:
All this time spent looking at this block of code has made me realize there's a substantial problem (more than just a wayward/inaccurate comment).  This block of code is calling method on "problem", without first checking that problem is not null.  It seems to me that this entire block of code belongs further up, inside the block around line 2500 that performs a not-null check for problem.  (Or else this block should re-perform the null check and provide the corresponding logging; I think it makes more sense to move it to the above place which is already doing "problem-related" processing>